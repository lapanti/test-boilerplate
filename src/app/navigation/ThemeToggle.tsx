import React, { useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'

import { isDarkModeSelector } from '../../store/ducks/theme/selectors'
import themeSlice from '../../store/ducks/theme/slice'

const Form = styled.form`
    justify-self: end;
    padding: 0 var(--m);
`

const Toggle = styled.label`
    --border-radius: calc(var(--l) / 2);

    position: relative;
    display: inline-block;
    cursor: pointer;
`

const HiddenInput = styled.input`
    display: none;
`

const Track = styled.div`
    position: relative;
    width: calc(var(--xl) + var(--xs));
    height: var(--xm);
    border-radius: var(--border-radius);
    background: var(--background-secondary);
    transition: background 0.25s ease;
`

const Moon = styled.span<{ visible: boolean }>`
    position: absolute;
    font-size: 14px;
    opacity: ${({ visible }) => (visible ? 1 : 0)};
    transition: opacity 0.25s ease;
`

const Sun = styled(Moon)`
    left: calc(var(--m) + var(--xs));
`

const Thumb = styled.button<{ isDarkMode: boolean }>`
    position: absolute;
    top: 0;
    cursor: pointer;
    background: white;
    height: var(--xm);
    width: var(--xm);
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.25);
    border-radius: var(--border-radius);
    border: 0;
    padding: 0;
    transition: transform 0.25s ease;
    transform: translateX(${({ isDarkMode }) => (isDarkMode ? 'calc(var(--m) + var(--xs))' : 0)});
`

const ThemeToggle: React.FunctionComponent = () => {
    const isDarkMode = useSelector(isDarkModeSelector)
    const dispatch = useDispatch()
    const toggle = useCallback(() => void dispatch(themeSlice.actions.toggleDarkMode()), [dispatch])
    return (
        <Form>
            <Toggle>
                <HiddenInput checked={isDarkMode} type="checkbox" onChange={toggle} />
                <Track>
                    <Moon visible={isDarkMode}>üåô</Moon>
                    <Sun visible={!isDarkMode}>‚òÄÔ∏è</Sun>
                </Track>
                <Thumb
                    aria-checked={isDarkMode ? 'true' : 'false'}
                    aria-label="Toggle dark/light mode"
                    isDarkMode={isDarkMode}
                    role="switch"
                    type="button"
                    onClick={toggle}
                />
            </Toggle>
        </Form>
    )
}

export default ThemeToggle
