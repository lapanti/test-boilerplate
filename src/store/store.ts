import { combineReducers, configureStore } from '@reduxjs/toolkit'

import themeSlice, { initialState as theme } from './ducks/theme/slice'

const reducer = combineReducers({ theme: themeSlice.reducer })

export type State = ReturnType<typeof reducer>

const store = configureStore({
    reducer,
})

export const mockStore = (state: Partial<State> = {}): typeof store =>
    configureStore({ preloadedState: { theme, ...state }, reducer })

export default store
