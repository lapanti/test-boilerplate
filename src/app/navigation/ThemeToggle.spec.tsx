import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { Provider } from 'react-redux'

import themeSlice from '../../store/ducks/theme/slice'
import { mockStore } from '../../store/store'
import ThemeToggle from './ThemeToggle'

jest.mock('../../lib/preference', () => ({
    prefersDarkMode: jest.fn().mockReturnValue(true),
}))

describe('<ThemeToggle />', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    it('should render correctly (dark mode)', () => {
        const store = mockStore({ theme: { darkMode: true } })
        const { container } = render(
            <Provider store={store}>
                <ThemeToggle />
            </Provider>
        )
        expect(screen.getByRole('switch', { name: /Toggle dark\/light mode/i })).toHaveAttribute('aria-checked', 'true')
        expect(container.firstChild).toMatchSnapshot()
    })

    it('should render correctly (light mode)', () => {
        const store = mockStore({ theme: { darkMode: false } })
        const { container } = render(
            <Provider store={store}>
                <ThemeToggle />
            </Provider>
        )
        expect(screen.getByRole('switch', { name: /Toggle dark\/light mode/i })).toHaveAttribute(
            'aria-checked',
            'false'
        )
        expect(container.firstChild).toMatchSnapshot()
    })

    it('should dispatch correct action on toggle', () => {
        const store = mockStore()
        jest.spyOn(store, 'dispatch')
        render(
            <Provider store={store}>
                <ThemeToggle />
            </Provider>
        )
        const toggle = screen.getByRole('switch', { name: /Toggle dark\/light mode/i })
        expect(toggle).toHaveAttribute('aria-checked', 'true')
        userEvent.click(toggle)
        expect(store.dispatch).toHaveBeenCalledWith(themeSlice.actions.toggleDarkMode())
        expect(toggle).toHaveAttribute('aria-checked', 'false')
    })
})
