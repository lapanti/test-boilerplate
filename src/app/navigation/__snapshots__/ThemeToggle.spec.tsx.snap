// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeToggle /> should render correctly (dark mode) 1`] = `
.c0 {
  justify-self: end;
  padding: 0 var(--m);
}

.c1 {
  --border-radius: calc(var(--l) / 2);
  position: relative;
  display: inline-block;
  cursor: pointer;
}

.c2 {
  display: none;
}

.c3 {
  position: relative;
  width: calc(var(--xl) + var(--xs));
  height: var(--xm);
  border-radius: var(--border-radius);
  background: var(--background-secondary);
  -webkit-transition: background 0.25s ease;
  transition: background 0.25s ease;
}

.c4 {
  position: absolute;
  font-size: 14px;
  opacity: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c5 {
  position: absolute;
  font-size: 14px;
  opacity: 0;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c6 {
  left: calc(var(--m) + var(--xs));
}

.c7 {
  position: absolute;
  top: 0;
  cursor: pointer;
  background: white;
  height: var(--xm);
  width: var(--xm);
  box-shadow: 0 0 10px rgba(0,0,0,0.25);
  border-radius: var(--border-radius);
  border: 0;
  padding: 0;
  -webkit-transition: -webkit-transform 0.25s ease;
  -webkit-transition: transform 0.25s ease;
  transition: transform 0.25s ease;
  -webkit-transform: translateX(calc(var(--m) + var(--xs)));
  -ms-transform: translateX(calc(var(--m) + var(--xs)));
  transform: translateX(calc(var(--m) + var(--xs)));
}

<form
  class="c0"
>
  <label
    class="c1"
  >
    <input
      checked=""
      class="c2"
      type="checkbox"
    />
    <div
      class="c3"
    >
      <span
        class="c4"
      >
        üåô
      </span>
      <span
        class="c5 c6"
      >
        ‚òÄÔ∏è
      </span>
    </div>
    <button
      aria-checked="true"
      aria-label="Toggle dark/light mode"
      class="c7"
      role="switch"
      type="button"
    />
  </label>
</form>
`;

exports[`<ThemeToggle /> should render correctly (light mode) 1`] = `
.c0 {
  justify-self: end;
  padding: 0 var(--m);
}

.c1 {
  --border-radius: calc(var(--l) / 2);
  position: relative;
  display: inline-block;
  cursor: pointer;
}

.c2 {
  display: none;
}

.c3 {
  position: relative;
  width: calc(var(--xl) + var(--xs));
  height: var(--xm);
  border-radius: var(--border-radius);
  background: var(--background-secondary);
  -webkit-transition: background 0.25s ease;
  transition: background 0.25s ease;
}

.c4 {
  position: absolute;
  font-size: 14px;
  opacity: 0;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c5 {
  position: absolute;
  font-size: 14px;
  opacity: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c6 {
  left: calc(var(--m) + var(--xs));
}

.c7 {
  position: absolute;
  top: 0;
  cursor: pointer;
  background: white;
  height: var(--xm);
  width: var(--xm);
  box-shadow: 0 0 10px rgba(0,0,0,0.25);
  border-radius: var(--border-radius);
  border: 0;
  padding: 0;
  -webkit-transition: -webkit-transform 0.25s ease;
  -webkit-transition: transform 0.25s ease;
  transition: transform 0.25s ease;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

<form
  class="c0"
>
  <label
    class="c1"
  >
    <input
      class="c2"
      type="checkbox"
    />
    <div
      class="c3"
    >
      <span
        class="c4"
      >
        üåô
      </span>
      <span
        class="c5 c6"
      >
        ‚òÄÔ∏è
      </span>
    </div>
    <button
      aria-checked="false"
      aria-label="Toggle dark/light mode"
      class="c7"
      role="switch"
      type="button"
    />
  </label>
</form>
`;
